pkg/gitconfig/config.go

```diff
@@ -0,0 +1,379 @@
+// to the vars map), update a key (key is the target key, value the new value)
+// or delete a key (parseFunc returns skip).
+func parseConfig(in io.Reader, key, value string, cb parseFunc) []string {
+	wSection, wSubsection, wKey := splitKey(key)
+
+	s := bufio.NewScanner(in)
+
+	lines := make([]string, 0, 128)
+	var section string
+	var subsection string
+	for s.Scan() {
+		line := s.Text()
+
+		lines = append(lines, line)
+
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+		if strings.HasPrefix(line, ";") {
+			continue
+		}
+		if strings.HasPrefix(line, "[") {
+			s, subs, skip := parseSectionHeader(line)
+			if skip {
+				continue
+			}
+			section = s
+			subsection = subs
+		}
+
+		if key != "" && (section != wSection && subsection != wSubsection) {
+			continue
+		}
+
+		kvp := strings.Split(line, "=")
+		trim(kvp)
+		if len(kvp) < 2 {
+			continue
+		}
+
+		fKey := section + "."
+		if subsection != "" {
+			fKey += subsection + "."
+		}
+		fKey += kvp[0]
+		if key == "" {
+			wKey = kvp[0]
+		}
+
+		oValue := kvp[1]
+		comment := ""
+
+		if strings.ContainsAny(oValue, "#;") {
+			comment = " " + oValue[strings.IndexAny(oValue, "#;"):]
+			oValue = oValue[:strings.IndexAny(oValue, "#;")]
+			oValue = strings.TrimSpace(oValue)
+		}
+
+		if key != "" && (key != fKey) {
+			continue
+		}
+		if key != "" {
+			oValue = value
+		}
+
+		newLine, skip := cb(fKey, wKey, oValue, comment)
+		if skip {
+			// remove the last line
+			lines = lines[:len(lines)-1]
+
+			continue
+		}
+		lines[len(lines)-1] = newLine
+	}
+
+	return lines
+}
+
+// NewFromMap allows creating a new preset config from a map.
+func NewFromMap(data map[string]string) *Config {
+	c := &Config{
+		readonly: true,
+		vars:     make(map[string]string, len(data)),
+	}
+
+	for k, v := range data {
+		c.vars[k] = v
+	}
+
+	return c
+}
+
+// LoadConfig tries to load a gitconfig from the given path.
+func LoadConfig(fn string) (*Config, error) {
+	fh, err := os.Open(fn)
+	if err != nil {
+		return nil, err
+	}
+	defer fh.Close() //nolint:errcheck
+
+	c := ParseConfig(fh)
+	c.path = fn
+
+	return c, nil
+}
+
+// ParseConfig will try to parse a gitconfig from the given io.Reader. It never fails.
+// Invalid configs will be silently rejceted.
+func ParseConfig(r io.Reader) *Config {
+	c := &Config{
+		vars: make(map[string]string, 42),
+	}
+
+	lines := parseConfig(r, "", "", func(fk, k, v, comment string) (string, bool) {
+		c.vars[fk] = v
+
+		return fmt.Sprintf(keyValueTpl, k, v, comment), false
+	})
+
+	c.raw.WriteString(strings.Join(lines, "\n"))
+	c.raw.WriteString("\n")
+
+	debug.Log("processed config: %s\nvars: %+v", c.raw.String(), c.vars)
+
+	return c
+}
+
+// LoadConfigFromEnv will try to parse an overlay config from the environment variables.
+// If no environment variables are set the resulting config will be valid but empty.
+// Either way it will not be writeable.
+func LoadConfigFromEnv(envPrefix string) *Config {
+	c := &Config{
+		noWrites: true,
```

Comment:
I feel like this is missing a `vars: make(map[string]string, 42)`

Reply:
```suggestion
		noWrites: true,
		vars: make(map[string]string, 42),
```

Reply:
Actually I realized afterwards you were doing things properly in Set and Get to avoid any issue with this, so feel free to disregard all of these.

Reply:
Ack.